### PR TITLE
[RLlib] Fix TQC critic divergence: stop actor-loss gradients from leaking into the critics

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
TQC's `_forward_train` computes the actor-loss Q-values (`qf_curr`) through the
online critics without the critic-parameter freeze SAC and CQL apply to this pass.
`SACTorchLearner.compute_gradients` (inherited by TQC) runs one backward pass per
optimizer -- in registration order `qf` -> `policy` -> `alpha` -- zeroing only that
optimizer's parameters and harvesting `param.grad` by reference. The policy backward
therefore accumulates the actor loss's critic gradients into the already-harvested
critic gradients, and the critic optimizer applies them: every update also pushes the
critics to raise Q-values at the policy's actions, bypassing the Bellman targets.

On hard envs this compounds: on `Humanoid-v4`, TQC's critic estimates grow without
bound while the episode return collapses, matching the long-standing "TQC doesn't
learn on Humanoid" reports (#61370). SAC is unaffected because its module freezes the
critic parameters for this exact pass ("straight-through": policy gradients flow
through the sampled actions, not the critic weights).

The fix is to apply SAC's freeze pattern to TQC's `qf_curr` forward pass. With the fix,
TQC on `Humanoid-v4` keeps its critic bounded and reaches SAC-level returns with
paper-default hyperparameters (n_critics=2, top_quantiles_to_drop_per_net=2); without
it, the critic diverges and the return collapses in the same-seed comparison.

### CODE REVIEW

**Comment by ArturNiederfahrenhorst** [comment]
Screenshot 2026-07-30 at 11 02 19 AM

Evidence of how this fixes q value overestimation

**Comment by ArturNiederfahrenhorst** [comment]
Screenshot 2026-07-30 at 11 03 25 AM

Evidence of this fix translating to higher rewards

File: rllib/algorithms/tqc/torch/default_tqc_torch_rl_module.py [hunk 1]
```diff
@@ -110,12 +109,24 @@ def _forward_train(self, batch: Dict[str, Any]) -> Dict[str, Any]:
...
         output[QF_PREDS] = qf_out  # (batch, n_critics, n_quantiles)
 
-        # Compute Q-values for resampled actions (for actor loss)
+        # Compute Q-values for resampled actions (for the actor loss).
+        # Freeze the critic parameters for this pass (same as SAC): the actor
+        # loss must only produce gradients w.r.t. the policy (through the
+        # actions), not w.r.t. the critics. Without this, the actor loss's
+        # critic gradients leak into the critic optimizer's update and push
+        # the critics to inflate Q-values at the policy's actions.
+        all_qf_params = list(self.qf_encoders.parameters()) + list(
```
Comment (pseudo-rnd-thoughts) [inline rllib/algorithms/tqc/torch/default_tqc_torch_rl_module.py:124]: Why can't we use `torch.no_grad()` also gemini's code above seems right

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
Overall looks good, thanks for fixing it. I just think that the no-grad stage needs to be revisited

File: rllib/algorithms/tqc/torch/default_tqc_torch_rl_module.py (hunk 1 above)
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/tqc/torch/default_tqc_torch_rl_module.py:124]: Re: torch.no_grad(), check out the SAC file!
It has the following comment at the same section of the forward function (what's where this fix is coming from!):

>         # Make sure we perform a "straight-through gradient" pass here,
>         # ignoring the gradients of the q-net, however, still recording
>         # the gradients of the policy net (which was used to rsample the actions used
>         # here). This is different from doing `.detach()` or `with torch.no_grads()`,
>         # as these two methds would fully block all gradient recordings, including
>         # the needed policy ones.

Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/tqc/torch/default_tqc_torch_rl_module.py:124]: But you are right, Gemini's comment makes sense I think

**Review by pseudo-rnd-thoughts — APPROVED** [review]
LGTM
